File: frontend/lib/src/features/reservation/data/dto/reservation_response_dto.dart
import "../../domain/entities/reservation_entity.dart";

class ReservationEventResponseDto {
  final String name;
  final String? occurredAt;

  const ReservationEventResponseDto({
    required this.name,
    required this.occurredAt,
  });

  factory ReservationEventResponseDto.fromJson(Map<String, dynamic> json) {
    return ReservationEventResponseDto(
      name: json["name"] as String,
      occurredAt: json["occurredAt"] as String?,
    );
  }

  ReservationDomainEventEntity toDomain() {
    return ReservationDomainEventEntity(
      name: name,
      occurredAt: occurredAt == null ? null : DateTime.parse(occurredAt!),
    );
  }
}

class ReservationResponseDto {
  final String id;
  final String guestId;
  final String roomId;
  final String status;
  final String checkIn;
  final String checkOut;
  final String createdAt;
  final String updatedAt;
  final List<ReservationEventResponseDto> domainEvents;

  const ReservationResponseDto({
    required this.id,
    required this.guestId,
    required this.roomId,
    required this.status,
    required this.checkIn,
    required this.checkOut,
    required this.createdAt,
    required this.updatedAt,
    required this.domainEvents,
  });

  factory ReservationResponseDto.fromJson(Map<String, dynamic> json) {
    final rawEvents = (json["domainEvents"] as List<dynamic>? ?? []);

    return ReservationResponseDto(
      id: json["id"] as String,
      guestId: json["guestId"] as String,
      roomId: json["roomId"] as String,
      status: json["status"] as String,
      checkIn: json["checkIn"] as String,
      checkOut: json["checkOut"] as String,
      createdAt: json["createdAt"] as String,
      updatedAt: json["updatedAt"] as String,
      domainEvents: rawEvents
          .map(
            (item) => ReservationEventResponseDto.fromJson(
                item as Map<String, dynamic>),
          )
          .toList(),
    );
  }

  ReservationEntity toDomain() {
    return ReservationEntity(
      id: id,
      guestId: guestId,
      roomId: roomId,
      status: status,
      checkIn: DateTime.parse(checkIn),
      checkOut: DateTime.parse(checkOut),
      createdAt: DateTime.parse(createdAt),
      updatedAt: DateTime.parse(updatedAt),
      domainEvents: domainEvents.map((event) => event.toDomain()).toList(),
    );
  }
}

File: src/modules/reservation/presentation/dto/reservation.response.dto.ts
import { Reservation } from "../../domain/aggregates/reservation/reservation.aggregate";

type ReservationEventDto = {
  name: string;
  occurredAt?: string;
};

export class ReservationResponseDto {
  id!: string;
  guestId!: string;
  roomId!: string;
  status!: string;
  checkIn!: string;
  checkOut!: string;
  createdAt!: string;
  updatedAt!: string;
  domainEvents!: ReservationEventDto[];

  static from(reservation: Reservation, domainEvents: unknown[] = []): ReservationResponseDto {
    const dto = new ReservationResponseDto();
    dto.id = reservation.id;
    dto.guestId = reservation.guestId;
    dto.roomId = reservation.roomId;
    dto.status = reservation.status;
    dto.checkIn = reservation.stayPeriod.checkIn.toISOString();
    dto.checkOut = reservation.stayPeriod.checkOut.toISOString();
    dto.createdAt = reservation.createdAt.toISOString();
    dto.updatedAt = reservation.updatedAt.toISOString();
    dto.domainEvents = domainEvents.map((event) => {
      const eventRecord = event as { eventName?: string; occurredAt?: Date };
      return {
        name: eventRecord.eventName ?? "UnknownDomainEvent",
        occurredAt: eventRecord.occurredAt?.toISOString(),
      };
    });
    return dto;
  }
}
